src/main/java/org/opensearch/ad/ratelimit/CheckpointWriteWorker.java

```diff
@@ -0,0 +1,299 @@
+import org.opensearch.ad.NodeStateManager;
+import org.opensearch.ad.breaker.ADCircuitBreakerService;
+import org.opensearch.ad.indices.AnomalyDetectionIndices;
+import org.opensearch.ad.ml.CheckpointDao;
+import org.opensearch.ad.ml.EntityModel;
+import org.opensearch.ad.ml.ModelState;
+import org.opensearch.ad.model.AnomalyDetector;
+import org.opensearch.ad.util.ClientUtil;
+import org.opensearch.ad.util.ExceptionUtil;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Strings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.threadpool.ThreadPool;
+
+public class CheckpointWriteWorker extends BatchWorker<CheckpointWriteRequest, BulkRequest, BulkResponse> {
+    private static final Logger LOG = LogManager.getLogger(CheckpointWriteWorker.class);
+
+    private final AnomalyDetectionIndices indexUtil;
+    private final CheckpointDao checkpoint;
+    private final String indexName;
+    private final Duration checkpointInterval;
+
+    public CheckpointWriteWorker(
+        long heapSizeInBytes,
+        int singleRequestSizeInBytes,
+        Setting<Float> maxHeapPercentForQueueSetting,
+        ClusterService clusterService,
+        Random random,
+        ADCircuitBreakerService adCircuitBreakerService,
+        ThreadPool threadPool,
+        Settings settings,
+        float maxQueuedTaskRatio,
+        Clock clock,
+        float mediumSegmentPruneRatio,
+        float lowSegmentPruneRatio,
+        int maintenanceFreqConstant,
+        ClientUtil clientUtil,
+        Duration executionTtl,
+        AnomalyDetectionIndices indexUtil,
+        CheckpointDao checkpoint,
+        String indexName,
+        Duration checkpointInterval,
+        NodeStateManager stateManager,
+        Duration stateTtl
+    ) {
+        super(
+            "checkpoint-write",
+            heapSizeInBytes,
+            singleRequestSizeInBytes,
+            maxHeapPercentForQueueSetting,
+            clusterService,
+            random,
+            adCircuitBreakerService,
+            threadPool,
+            settings,
+            maxQueuedTaskRatio,
+            clock,
+            mediumSegmentPruneRatio,
+            lowSegmentPruneRatio,
+            maintenanceFreqConstant,
+            clientUtil,
+            CHECKPOINT_WRITE_QUEUE_CONCURRENCY,
+            executionTtl,
+            CHECKPOINT_WRITE_QUEUE_BATCH_SIZE,
+            stateTtl,
+            stateManager
+        );
+        this.indexUtil = indexUtil;
+        this.checkpoint = checkpoint;
+        this.indexName = indexName;
+        this.checkpointInterval = checkpointInterval;
+    }
+
+    @Override
+    protected void executeBatchRequest(BulkRequest request, ActionListener<BulkResponse> listener) {
+        if (indexUtil.doesCheckpointIndexExist()) {
+            clientUtil.<BulkRequest, BulkResponse>execute(BulkAction.INSTANCE, request, listener);
```

Comment:
Why doesnt this use the checkpointDao for putting checkpoints? Shouldnt all calls to put and get checkpoints go through the checkpointDao?

Reply:
This line is a bulk call.  It does not involve any specifics of checkpoint.  CheckpointDao encapsulates inner working mechanisms of checkpoint index. Also, creating a one-line function inside CheckpointDao is wasteful.  So it seems fine to put this line here.

Reply:
Same discussion happening in above comment. Resolving